### PR TITLE
Rpcv03 starknet trace block transactions

### LINKED ISSUE #315: starknet_traceBlockTransactions
implement https://github.com/starkware-libs/starknet-specs/blob/v0.3.0/api/starknet_trace_api_openrpc.json#L111

### PR DESCRIPTION
**Pull request by rianhughes**
closes https://github.com/NethermindEth/starknet.go/issues/315

### CODE REVIEW

File: rpc/mock_test.go [hunk 1]
```diff
@@ -623,3 +626,41 @@ func mock_starknet_getBlockWithTxHashes(result interface{}, method string, args
...
+			return err
+		}
+		BlockTrace, err := json.Marshal(rawBlockTrace.Result)
+		if err != nil {
+			return err
+		}
+		if nil != json.Unmarshal(BlockTrace, &r) {
+			return err
+		}
+		return nil
```
Comment (kirugan) [inline rpc/mock_test.go]: let's simplify to `return json.Unmarshal(BLockTrace, &r)`

File: rpc/provider.go [hunk 2]
```diff
@@ -48,6 +48,7 @@ type api interface {
 	AddDeclareTransaction(ctx context.Context, declareTransaction BroadcastedDeclareTransaction) (*AddDeclareTransactionResponse, error)
 	AddDeployAccountTransaction(ctx context.Context, deployAccountTransaction BroadcastedDeployAccountTransaction) (*AddDeployAccountTransactionResponse, error)
 	SimulateTransactions(ctx context.Context, blockID BlockID, txns []BroadcastedTransaction, simFlags []SimulationFlag) ([]SimulatedTransaction, error)
+	TraceBlockTransactions(ctx context.Context, blockHash *felt.Felt) (*[]Trace, error)
```
Comment (kirugan) [inline rpc/provider.go]: is it possible to return just `[]Trace` not a pointer to slice?

File: rpc/mock_test.go [hunk 3]
```diff
@@ -623,3 +626,41 @@ func mock_starknet_getBlockWithTxHashes(result interface{}, method string, args
...
+	if !ok || r == nil {
+		return errWrongType
+	}
+	if len(args) != 1 {
+		return errWrongArgs
+	}
+	blockHash, ok := args[0].(*felt.Felt)
+	if !ok {
+		fmt.Printf("args[0] should be felt, got %T\n", args[0])
+		return errWrongArgs
```
Comment (kirugan) [inline rpc/mock_test.go]: wdty about wrapping this error i.e. errors.Wrap(err, "...msg above..."

File: rpc/provider.go (hunk 2 above)
Comment (rianhughes) [reply rpc/provider.go]: The traces can be quite large (eg a few MB for mainnet), so I thought it would be more efficient to return a pointer. Wdyt?

Comment (rianhughes) [reply rpc/provider.go]: Leaving a note for any future readers, Go slices don't copy the entire data because internally they use pointers.

**Review by cicr99 — APPROVED** [review]
lgtm

**Pull request merged**